File: lower/lower.py
import string
import random
import time

from flask import Flask, Response
from flask.json import jsonify
import requests

from opentelemetry import trace
from opentelemetry.trace import StatusCode, Status
from opentelemetry.instrumentation.flask import FlaskInstrumentor
from opentelemetry.instrumentation.requests import RequestsInstrumentor
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter

app = Flask(__name__)

trace.set_tracer_provider(TracerProvider(resource=Resource.create({"service.name": "lower"})))
span_exporter = OTLPSpanExporter(endpoint="collector:4317")
trace_provider = trace.get_tracer_provider()
trace_provider.add_span_processor(BatchSpanProcessor(span_exporter))
FlaskInstrumentor().instrument_app(app, tracer_provider=trace_provider)
RequestsInstrumentor().instrument()
tracer = trace.get_tracer(__name__)


def work(mu: float, sigma: float) -> None:
    # simulate work being done
    time.sleep(max(0.0, random.normalvariate(mu, sigma)))


def get_digit() -> int:
    with tracer.start_as_current_span("get_digit") as span:
        try:
            response = requests.get("http://digit:5000/")
            c = response.json()['char']
            return int(c)
        except Exception as e:
            e = Exception(f"FAILED to fetch a digit char")
            span.record_exception(e)
            span.set_status(Status(StatusCode.ERROR, str(e)))
            raise e


def random_lower() -> str:
    with tracer.start_as_current_span("random_lower") as span:
        _ = get_digit()
        c = random.choice(string.ascii_lowercase)
        span.set_attribute('char', c)
        return c


def process_lower(c: str) -> str:
    with tracer.start_as_current_span("process_lower") as span:
        span.set_attribute('char', c)
        span.add_event("processing lower char", {'char': c})
        work(0.0001, 0.00005)

        # 1/100 calls is extra slow
        if random.random() > 0.99:
            span.add_event("extra work", {'char': c})
            work(0.0002, 0.0001)
        
        # these chars are extra slow
        if c in {'Z', 'X', 'R',}:
            with tracer.start_as_current_span(f"extra_process_lower") as span:
                span.set_attribute('char', c)
                work(0.005, 0.0005)
        
        # these chars are extra slow too
        if c in {'Z', 'A', 'T'}:
            with tracer.start_as_current_span(f"extra_extra_process_lower") as span:
                span.set_attribute('char', c)
                work(0.0001, 0.00008)
        return c


def render_lower(c: str) -> Response:
    with tracer.start_as_current_span(f"render_lower") as span:
        span.set_attribute('char', c)
        work(0.0002, 0.0001)
        return jsonify(char=c)


@app.route('/')
def lower():
    c = random_lower()
    c = process_lower(c)
    return render_lower(c)


if __name__ == '__main__':
    app.run()

File: lower/lower.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Copyright The OpenTelemetry Authors
#
# SPDX-License-Identifier: Apache-2.0
require 'rubygems'
require 'bundler/setup'
require 'faraday'
require 'opentelemetry/sdk'
require 'sinatra/base'
require 'json'
require 'net/http'

Bundler.require

ENV['OTEL_TRACES_EXPORTER'] ||= 'otlp'
ENV['OTEL_PROPAGATORS'] ||= 'tracecontext,baggage,b3'
ENV['OTEL_EXPORTER_OTLP_ENDPOINT'] ||= 'http://collector:4318'

OpenTelemetry::SDK.configure do |c|
   c.service_name = "lower"
   c.logger.level = Logger::DEBUG
   c.logger.debug("Using OTLP endpoint: #{ENV['OTEL_EXPORTER_OTLP_ENDPOINT']}")
   c.use_all
end


def sinatra_tracer
  OpenTelemetry.tracer_provider.tracer('sinatra', '1.0')
end

# Rack middleware to extract span context, create child span, and add
# attributes/events to the span
class OpenTelemetryMiddleware
  def initialize(app)
    @app = app
    @tracer = sinatra_tracer
  end

  def call(env)
    # Extract context from request headers
    context = OpenTelemetry.propagation.extract(
      env,
      getter: OpenTelemetry::Common::Propagation.rack_env_getter
    )

    status, headers, response_body = 200, {}, ''
    #OpenTelemetry.logger.debug("One more request #{env.inspect}")

    # Span name SHOULD be set to route:
    span_name = env['PATH_INFO']

    # For attribute naming, see
    # https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-semantic-conventions.md#http-server

    # Activate the extracted context
    OpenTelemetry::Context.with_current(context) do
      # Span kind MUST be `:server` for a HTTP server span
      @tracer.in_span(
        span_name,
        attributes: {
          'component' => 'http',
          'http.method' => env['REQUEST_METHOD'],
          'http.route' => env['PATH_INFO'],
          'http.url' => env['REQUEST_URI'],
        },
        kind: :server
      ) do |span|
        # Run application stack
        status, headers, response_body = @app.call(env)

        span.set_attribute('http.status_code', status)
      end
    end

    [status, headers, response_body]
  end
end

set :bind, '0.0.0.0'
set :port, 5000

use OpenTelemetryMiddleware
CHARS = ('a'..'z').to_a
def work time, span_name, c
  sinatra_tracer.in_span(span_name, attributes: { "char" => c}, kind: :server) do |span|
    sleep time
  end
end

def get_digit(c)
  sinatra_tracer.in_span("get_digit", attributes: { "char" => c}, kind: :server) do |span|
    begin
      digit = Net::HTTP.get(URI("http://digit:5000"))
      json = JSON.parse(digit)
      json['char']
    rescue Error => e
      span.record_exception(e)
      span.status = OpenTelemetry::Trace::Status.error
    end
  end
end

def prepare_char
  c = CHARS.sample
  get_digit(c)
  case c
  when 'z', 'x', 'r'
    work(0.05, 'extra_process_lower',  c)
  when 'z', 'a', 't'
    work(0.01, 'extra_extra_process_lower',  c)
  end
  c
end

get '/' do
  content_type :json

  c = prepare_char

  {char: c}.to_json
end
